### ISSUE TITLE
Tracking issue for RFC 2603, "Symbol Mangling v2"

### ISSUE THREAD

**Issue by Centril** [rust-lang/rust#60705]
This is a tracking issue for the RFC "Symbol Mangling v2" (rust-lang/rfcs#2603).

**Steps:**

- [ ] Implement the RFC (cc @rust-lang/compiler -- can anyone write up mentoring instructions?)
- [ ] Adjust documentation ([see instructions on rustc-guide][doc-guide])
- [ ] Stabilization PR ([see instructions on rustc-guide][stabilization-guide])

[stabilization-guide]: https://rust-lang.github.io/rustc-guide/stabilization_guide.html#stabilization-pr
[doc-guide]: https://rust-lang.github.io/rustc-guide/stabilization_guide.html#documentation-prs

**Unresolved questions:**

- [ ] [Punycode vs UTF-8](https://github.com/rust-lang/rfcs/blob/master/text/2603-symbol-name-mangling-v2.md#punycode-vs-utf-8)
- [ ] [Encoding parameter types for function symbols](https://github.com/rust-lang/rfcs/blob/master/text/2603-symbol-name-mangling-v2.md#encoding-parameter-types-for-function-symbols)

**Comment by eddyb** [comment]
cc @rust-lang/compiler 

I'd add this as an unresolved question: https://github.com/rust-lang/rfcs/pull/2603#issuecomment-489649554

I'd rather change the grammar now, if ever, before merging https://github.com/rust-lang/rust/pull/57967.

**Comment by michaelwoerister** [comment]
I think the `Punycode vs UTF-8` and `Encoding parameters types for function symbols` questions are actually resolved. The RFC text contains a recommendation on what the resolution should look like (since before the motion to merge) and nobody objected.

**Comment by eddyb** [comment]
I tried the "type/value namespace" shorthands, and this is what I got, stats-wise:
(apologies for having two tables, but it was hard to combine them further)

### 986807 symbols: sizes and compression ratios (▶)
#### `v0` without shorthands (types use `Nt` and values use `Nv`)
|  | legacy | mw | ▶ | mw + compress | v0 | ▶ | v0 + compress |
| ---: | --- | --- | --- | --- | --- | --- | --- |
| min | 30 | 25 | 0.97x | 25 | 25 | 0.96x | 25 |
| avg | 81 | 313 | 1.46x | 193 | 318 | 1.47x | 194 |
| max | 688 | 13409 | 14.83x | 1115 | 13759 | 15.24x | 1105 |
| total | 76.6MiB | 295.4MiB | 1.62x | 182.0MiB | 300.2MiB | 1.64x | 183.5MiB |

#### `v0` with shorthands (types use `Z` and values use `V`)
|  | legacy | mw | ▶ | mw + compress | v0 | ▶ | v0 + compress |
| ---: | --- | --- | --- | --- | --- | --- | --- |
| min | 30 | 25 | 0.97x | 25 | 24 | 0.96x | 24 |
| avg | 81 | 313 | 1.46x | 193 | 303 | 1.48x | 184 |
| max | 688 | 13409 | 14.83x | 1115 | 12925 | 14.99x | 1047 |
| total | 76.6MiB | 295.4MiB | 1.62x | 182.0MiB | 285.5MiB | 1.64x | 173.9MiB |

So that's around 5% reduction, kinda nice, but not that much.

**Comment by robinmoussu** [comment]
I just opened a bug in [compiler-explorer](https://github.com/mattgodbolt/compiler-explorer/issues/1754#issuecomment-569268520) because I noticed collision in the demangled name of different monomorphisation of the same function. As far as I understand, the current mangling scheme of rust still use the v1 witch doesn't contains the required information, while this proposed v2 would solved it? Did I correctly understood what the current situation is?

**Comment by eddyb** [comment]
@robinmoussu As per #57967, you can set `RUSTFLAGS=-Zsymbol-mangling-version=v0` on nightly.

**Comment by crlf0710** [comment]
@eddyb @michaelwoerister 
I wonder if it's possible/feasible to put some "mark" into the mangled symbol name to indicate that this stack frame is preferred not to disable in backtraces, #68336.

**Comment by eddyb** [comment]
Could just rely on a `_` prefix or something similar, but either way I would prefer not to add something that marginally related, to the symbol mangling syntax.

**Comment by eddyb** [comment]
Updating the RFC's "v2" to "v0" as per https://github.com/rust-lang/rust/pull/57967#issuecomment-463596992 (should've done it back then).

**Comment by eddyb** [comment]
I've just submitted the last patch, "[[PATCH] Support the new ("v0") mangling scheme in rust-demangle.](https://gcc.gnu.org/pipermail/gcc-patches/2020-March/542012.html)", to the `gcc-patches` mailing list (yes, the demangler used by `binutils` and `gdb` is in `libiberty`, the main copy of which is in GCC - would've been much easier to contribute to it otherwise, *oh well*).

If everything goes well, once that's merged we'll be able to also upstream the same implementation to `lldb`, Linux `perf`, `valgrind`, etc. - that's been the main blocker for rolling out the `v0` mangling.

**Comment by eddyb** [comment]
Potential change we might want to make to const value mangling: https://github.com/rust-lang/rust/issues/61486#issuecomment-599085848.

The tricky aspect is whether we change the way `_: usize` is encoded today, from `jp` to `p`, or we leave it as-is so that demanglers can keep working unchanged.

**Comment by eddyb** [comment]
Quick update: I haven't received a response [on the final patch](https://gcc.gnu.org/pipermail/gcc-patches/2020-March/542012.html) in 2.5 months.

This means I can probably resubmit with the placeholder constant syntax removed (see my previous comment, i.e. https://github.com/rust-lang/rust/issues/60705#issuecomment-599087150), and/or maybe we can even implement some form of ADT mangling in the interim, since it doesn't seem to be that difficult (see https://github.com/rust-lang/rust/pull/71232 for something similar).

But this is disappointing and another delay in getting the new mangling scheme to be supported in host tooling, which would've ideally been all dealt with last year.

**Comment by mark-i-m** [comment]
Perhaps an MCP could be used to get it on the compiler team radar and push it through?

**Comment by eddyb** [comment]
@mark-i-m What do you mean? A Rust MCP won't make GCC devs suddenly respond to my emails.

**Comment by mark-i-m** [comment]
Oh wait it won't 🤯 

Lol, sorry, I misunderstood.

**Comment by bstrie** [comment]
So support for this in rustc is completely implemented at this point, and now we're just waiting for tooling vendors to accept patches? If so, rather than waiting passively does anyone have contacts in the GCC project who we can reach out to to shepherd this along?

**Comment by benesch** [comment]
I've asked Ian Taylor to look at it, since he's reviewed many of the last patches to the demangler: https://gcc.gnu.org/pipermail/gcc-patches/2020-October/557390.html If Ian is busy, perhaps we can ask Jeff Law or Jakub Jelinek, who have committed the prior patches.

**Comment by eddyb** [comment]
@benesch's [updated version of the patch](https://gcc.gnu.org/pipermail/gcc-patches/2020-November/558905.html) just landed :tada:

It includes ports of @varkor's 3 PRs related to const generics:
* https://github.com/alexcrichton/rustc-demangle/pull/39 "Add demangling for `bool`, signed integers, and `char`"
* https://github.com/alexcrichton/rustc-demangle/pull/40 "Elide the type when the const value is a placeholder `p`"
* https://github.com/alexcrichton/rustc-demangle/pull/41 "Print `char` using `{:?}`"

Thanks to @benesch for updating the patch, pushing it through, and for @jwakely for committing it!

**Comment by benesch** [comment]
And thank _you_, @eddyb, for doing all the heavy lifting here! I'll be very excited when we can stabilize the new mangling scheme.

**Comment by bstrie** [comment]
eddyb previously wrote: 

> If everything goes well, once that's merged we'll be able to also upstream the same implementation to lldb, Linux perf, valgrind, etc. - that's been the main blocker for rolling out the v0 mangling.

So is that the next step? I'm unsure if by "upstream" that means that the same code will need to be landed in all those projects manually, or if it's implying that those are all using libiberty and we'll just need to wait for them to upgrade their dependency.

**Comment by Amanieu** [comment]
In the hopes of moving forward with this, I propose the following plan:
- Backport the libiberty demangler changes to the gcc-8, gcc-9 and gcc-10 branches. This should cover most actively used distros including debian 10 (gcc-8) and ubuntu 20.04 (gcc-9).
- No changes to perf are needed: it will automatically pick up the new demangler in libiberty.
- Valgrind uses a vendored copy of libiberty for demangling, I am working on a patch to update it to the latest version from GCC.
- I'm less sure about LLDB since it uses a completely different demangler implementation written in C++. I think we can skip this for now since LLDB didn't support the old mangling, so switching to v0 isn't really a regression here.

I think that we can switch the default name mangling to v0 on nightly once the GCC patches have been backported and distros have updated their GCC package.

**Comment by Amanieu** [comment]
Valgrind patch: https://bugs.kde.org/show_bug.cgi?id=431306

**Comment by ojeda** [comment]
binutils/gdb also were waiting for GCC's libiberty but it was done three days ago: https://github.com/bminor/binutils-gdb/commit/d750c713c9a34c8835e8e60370708cae675edb40

If we are backporting it to gcc-8 for distros etc., I guess those need backporting too.

**Comment by Amanieu** [comment]
Hmm that's inconvenient, I was expecting binutils to use gcc's libiberty rather than vendoring a copy. Considering the situation, I think it's only worth backporting the demangler to the latest stable releases which are:
- GCC 10
- binutils 2.35
- GDB 10

Distros maintaining older versions will just have to backport the changes themselves.

**Comment by Amanieu** [comment]
v0 name mangling support has been added to the GCC 10 development branch: https://sourceware.org/bugzilla/show_bug.cgi?id=27194

I think we should switch the default name mangling to v0 on nightly once GDB 10.2 is released.

**Comment by bstrie** [comment]
Before stabilization, do there exist benchmarks to determine if this has an effect has on compilation times? Presumably longer symbol names would have a negative effect on performance, so do we have an idea of how the length of an average symbol compares between the new and old schemes?

**Comment by bjorn3** [comment]
I couldn't fine any past perf run. Even the original implementation PR seems to miss one. I opened https://github.com/rust-lang/rust/pull/81137 for a perf run.

**Comment by bstrie** [comment]
To get some quick points of comparison, I built regex and ripgrep using both schemes.

Compilation time for regex:
```
$ RUSTFLAGS= time cargo build
    Finished dev [unoptimized + debuginfo] target(s) in 8.65s
        8.78 real        25.07 user         1.52 sys
$ RUSTFLAGS=-Zsymbol-mangling-version=v0 time cargo build
    Finished dev [unoptimized + debuginfo] target(s) in 8.88s
        9.01 real        26.09 user         1.57 sys
```

Symbol size for regex:
```
$ nm libregex-old.rlib | grep -o "__Z.*$" | wc -c
822521
$ nm libregex-new.rlib | grep -o "__R.*$" | wc -c
1156947
```

Compilation time for ripgrep:
```
$ RUSTFLAGS= time cargo build
    Finished dev [unoptimized + debuginfo] target(s) in 36.67s
       36.80 real       176.98 user        11.27 sys
$ RUSTFLAGS=-Zsymbol-mangling-version=v0 time cargo build
    Finished dev [unoptimized + debuginfo] target(s) in 37.67s
       37.80 real       178.71 user        11.40 sys
```

Symbol size for ripgrep:
```
$ nm target/debug/rg-old | grep -o "__Z.*$" | wc -c
3298005
$ nm target/debug/rg-new | grep -o "__R.*$" | wc -c
5415482
```

So regex sees its symbols grow by 40%, and ripgrep's symbols grow by 60%. Both take 2% longer to compile in a single trial, which is within the margin of error but worth keeping in mind.

**Comment by bstrie** [comment]
Looks like bjorn's perf build is ICEing in the test suite due to panics in rustc_symbol_mangling, which suggests this could use some battle-testing before stabilization. Would it be uncontroversial to make v0 the default in nightly only, without letting that default get uplifted into beta until formal stabilization has occurred?

**Comment by Mark-Simulacrum** [comment]
We can switch it on for (some of) our test builders, but I am opposed to enabling on nightly prior to at least a crater run and our tests passing.

**Comment by estebank** [comment]
> Would it be uncontroversial to make v0 the default in nightly only

I've seen cases in the past of bugs not being caught until beta (or even worse, stable) because nightly wasn't using the stable algorithm. If we were going to do something along those lines I would prefer if it was only in a single platform or some other similar strategy to avoid us finding that we've introduced a bug later in the release train and have to scramble to fix it.

**Comment by Amanieu** [comment]
I would prefer holding back on changing the default on nightly until we at least have support for v0 in a released version of GDB.

**Comment by bjorn3** [comment]
The perf run indicated significant regressions up to 10%.

**Comment by Amanieu** [comment]
It seems that almost all of the the runtime increase comes from `LLVM_module_codegen_emit_obj` and `run_linker`, which explains why the % increase is so high on incremental builds.

Also keep in mind that the % are regressions in instruction count, which has less impact on the wall time than it may seem since the linker is multi-threaded.

**Comment by michaelwoerister** [comment]
Those perf results are rather interesting. I wish we had less noisy wall-time data.

I think *building* the symbol names could be made faster by introducing some kind of caching, but the perf data seems to suggest that the slowdown actually comes from feeding longer symbol names to the linker. @eddyb collected some data a while ago, comparing symbol name lengths with various schemes: https://gist.github.com/eddyb/786598131525ef5adc9189a30e31c2fc

The "legacy" encoding (i.e. the current default) is indeed more compact because it can afford to throw away information about generic arguments. I think some increase in symbol name length is inevitable once we don't use the fixed size disambiguation hashes anymore. So I think we'll just have to bite the bullet in that respect.

**Comment by lnicola** [comment]
Switching to lld (which is planned, but currently in limbo, I guess) might also mitigate this. From what I've seen it's much faster on large amounts of debug info.

**Comment by michaelwoerister** [comment]
I checked the `lldb` box in the TODO list above. LLDB does not seem to do any demangling of Rust symbols so there is nothing to do there.

**Comment by michaelwoerister** [comment]
I've added resolving https://github.com/rust-lang/rust/issues/83611 to the TODO list. I'm looking forward to collecting more regression tests as we progress `:)`

**Comment by marmeladema** [comment]
GDB 10.2 has been released with the new rust demangler: https://www.gnu.org/software/gdb/news/

Does this mean we can start the crater run?

**Comment by michaelwoerister** [comment]
We might want to get https://github.com/rust-lang/rust/issues/83611 before that. @camelid has opened https://github.com/rust-lang/rust/pull/83767 to that end but I'm not quite sure what the state of that PR is.

**Comment by camelid** [comment]
> We might want to get #83611 before that. @camelid has opened #83767 to that end but I'm not quite sure what the state of that PR is.

Yeah, I've been meaning to work on that again. I'll hopefully be able to get to that soon.

**Comment by michaelwoerister** [comment]
#83767 has been merged now, thanks @camelid!

Before we do a crater run, can we make sure to add an assertion to the symbol mangling code in the compiler that checks that all symbol names generated can be demangled by the `rustc-demangle` crate? That would make the crater run much more useful.

**Comment by asm89** [comment]
Looks like GitHub picked up a reference to this issue in the commit, but LLDB trunk now supports Rust's symbol mangling scheme (v0). @tmiasko has been working on support for the mangling scheme in LLVM's Demangle library. It's now used by LLDB.

**Comment by bstrie** [comment]
If v0 becomes the default, will it also become the default for the stdlib? The argument in favor of making this the default assumes that anyone who doesn't want it will be able to turn it off, but you won't be able to "turn it off" for std unless you recompile std yourself.

**Comment by bjorn3** [comment]
Why would you not want it? I would assume that the old symbol mangling scheme will be removed somewhere after the switch.

**Comment by bstrie** [comment]
By "not want it" I just mean that there might exist people who are using tools that only understand the legacy mangling. I see no reason why anyone would want the legacy mangling on principle, but perhaps there's something preventing them from upgrading to newer versions of their tools? The question is, what is the reasonable time to wait in order to accommodate such users and give them time to upgrade? Nobody wants to delay this feature any longer than it already has been, but it's worth acknowledging that unless libstd has ships two copies of its symbols somehow, making v0 the default might cause problems for users on older tools.

**Comment by michaelwoerister** [comment]
@tmiasko has done a great job at summarizing the results of the crate run here: https://github.com/rust-lang/rust/pull/85530#issuecomment-857855379

It looks like const generics are the main blocker. @eddyb (or anyone else), do you have ideas about how to proceed there? Would it make sense to have generic support for unstable things in the mangling scheme that allows to opaquely encode things for which we don't have an official grammar yet?

**Comment by eddyb** [comment]
We could introduce something like `p` but which instead of being `_` takes a hash, so we can opaquely encode types/constants/etc. that we don't currently support mangling, as their "stable hash".

Other than that, we need to actually design a structural mangling scheme for constants, and if we have that we can "just" make it official, I think.

**Comment by eddyb** [comment]
@bstrie keep in mind that changing the mangling scheme is as easy as using `-Z build-std` - in fact, I've recommended using `env RUSTFLAGS=-Zsymbol-mangling-version=v0 cargo build -Z build-std` in the past, in order to get a fully-`v0` build. The toggle likely won't be on stable, though, unless there's at least active demand for it (but that is, of course, no guarantee by itself).

**Comment by michaelwoerister** [comment]
OK, I think it's clear that we don't want to block the mangling scheme on const generics reaching their final state, right? 

I see three possibilities:
1. Just use the existing `<const> = <type> <const-data>` production and rely on the `<type>` to give enough information to demanglers for deciding if `<const-data>` is a hash value. The comment on `<const-data>` would then be:
   ```
   // The encoding of a constant depends on its type. Integers use their value,
   // in base 16 (0-9a-f), not their memory representation. Negative integer
   // values are preceded with "n". The bool value false is encoded as `0_`, true
   // value as `1_`. The char constants are encoded using their Unicode scalar
   // value.
   // <-- new text starting here -->
   // For any other type the base-16 number is an opaque hash of the constant's 
   // value and cannot be demangled. Future versions of this mangling scheme
   // will likely define a structural encoding of constants that does enable
   // demangling complex values.
   <const-data> = ["n"] {<hex-digit>} "_"
   ```
2. Add another alternative production to the `<const>` production to make this more explicit:
   ```
   <const> = <type> <const-data>
           | "H" {<hex-digit>} "_" // opaque hash of const value
           | "p" // placeholder, shown as _
           | <backref>
   ```
3. Add a more general way of encoding things that demanglers can safely ignore. 

Option (1) seems the least disruptive. Is there a reason we can't use it to bridge the time until we have a proper encoding? I also imagine that we'll encounter const values that are just too big/complex to fully encode, so having a standard way of encoding things as stable hashes seems like a good idea in any case, right?

**Comment by eddyb** [comment]
Honestly, I was going to offer to take a stab at defining the mangling for ADT constructors ("2." in https://github.com/rust-lang/rust/issues/61486#issuecomment-878932102), it doesn't seem much harder than adding an opaque hash-only leaf (if we do that, I would prefer it to also be available for types too, FWIW).

And on the mangler side, we do have `destructure_const` to use today to take apart a non-leaf constant (we already use it for printing, which legacy mangling can trigger IIRC?).

*However,* @oli-obk asked me to wait on valtrees (which aren't *required* AFAIK, except perhaps to ensure we can't ever get duplicate symbols because of miri representations... which would also be a problem with stable hashes, right?).

> Option (1) seems the least disruptive. Is there a reason we can't use it to bridge the time until we have a proper encoding?

It doesn't really fit in the design I'm imagining, which would reuse e.g. the array type syntax (`"A" <type>`) to represent array values (`"A" <const>* "E"`), as mentioned in https://github.com/rust-lang/rust/issues/61486#issuecomment-830017281 - but also, demanglers don't *already* support it, meaning it's as much effort as having a special prefix for opaque hash-only leaves, *but also* it eats up most of the encoding space for no good reason.

**Comment by oli-obk** [comment]
yea... just make sure that anything that uses `destructure_const` needs a feature gate, I'm not too sure it behaves soundly for const generics in all cases

**Comment by michaelwoerister** [comment]
@eddyb, I'm not sure we are talking about the same thing here. I'm proposing to make `rustc` emit `<type> <const-data>` (with `<const-data>` being a hash of the constants value) just as a temporary solution until we have a proper grammar for ADT constants. Are you proposing to get rid of the `<const> = <type> <const-data>` production entirely?

**Comment by eddyb** [comment]
> Are you proposing to get rid of the `<const> = <type> <const-data>` production entirely?

That production is ~~a lie~~ more general than what is implemented (which is only integer types, `bool` and `char` for the `<type>` part).

Right now demanglers will treat any other type in that position as an error, so using it for that purpose requires changing demanglers, just like adding a special form for the opaque hash-only leaves, but leaving less encoding space usable by any future ADT mangling.

**Comment by michaelwoerister** [comment]
@eddyb: OK, I'm all for solving this properly now -- iff we can get it done in a reasonable amount of time 😃 

It sounds like you and @oli-obk have come up with an [exhaustive list](https://github.com/rust-lang/rust/issues/61486#issuecomment-878932102) of things the grammar needs to support, right? But it also sounds like the implementation on the compiler side is somewhat complicated by the fact that `destructure_const` isn't quite reliable yet?

**Comment by oli-obk** [comment]
>  But it also sounds like the implementation on the compiler side is somewhat complicated by the fact that destructure_const isn't quite reliable yet?

it should work just fine for all normal aggregates like arrays, tuples, enums and structs, but there are certainly types it cannot handle or will handle weirdly. I'm fairly certain that all currently legal const generic types will work just fine, so that should be ok. It's just that extending the list of legal types is not trivially ok.

**Comment by eddyb** [comment]
> @eddyb: OK, I'm all for solving this properly now -- iff we can get it done in a reasonable amount of time

Initial mangling implementation (and grammar) up at #87194 - I spent more time convincing myself that I couldn't cut certain corners in the grammar, writing the comments for it, and refactoring the current handling of placeholders (which is in a separate commit), than adding the new support.

That's mostly because `deref_const` and `destructure_const` already exist, and have been around for a few months if not longer, so we could've had this done before #85530 was opened - a lot of my PR is just copy-paste from `ty::print::pretty`, and adjusting the output to be the mangling we want (instead of user-facing).

Hopefully the extended constant mangling grammar doesn't end up being a bikeshed of its own.

**Comment by joshtriplett** [comment]
@eddyb Right, I was proposing stabilizing the option but *not* guaranteeing that it supports any particular value. We could choose to stabilize the `v0` value now, and then *consider* stabilizing the `legacy` value when we change the default (since there's no reason to pass it at all until the default changes).

**Comment by nnethercote** [comment]
An issue came up in https://github.com/rust-lang/rust/pull/89917#issuecomment-963755731 that is worth mentioning here: the compiler currently generates some v0 symbols that have a `.llvm.<numbers>` suffix, which violate the v0 spec (it doesn't allow '.' chars), and some v0 demangler implementations (libiberty and Valgrind) fail to demangle these symbols.

Either the compiler should be fixed to not append these suffixes (which may be hard, because it's LLVM that's adding them) or the v0 spec should be modified to permit these suffixes, and the libiberty/Valgrind implementations should be updated accordingly.

**Comment by eddyb** [comment]
> Either the compiler should be fixed to not append these suffixes (which may be hard, because it's LLVM that's adding them) or the v0 spec should be modified to permit these suffixes, and the libiberty/Valgrind implementations should be updated accordingly.

I don't think either of them is correct - or at least `rustc-demangle` doesn't do either, and does handle those pesky suffixes.

Does C++ Itanium mangling allow for the `.llvm.` suffixes? AFAIK no, but you should get them if you use Clang with LTO.

What `rustc-demangle` does, and what these tools should probably also do, is limit the symbol to just before the suffix, *before* attempting to demangle at all with *any* mangling scheme.

IOW, I believe this current behavior is a bug:
```console
$ c++filt _ZN3foo3barE
foo::bar
$ c++filt _ZN3foo3barE.llvm.123
_ZN3foo3barE.llvm.123
```
(at least in a world where LLVM's LTO exists - one could argue that they screwed up by doing this)

**Comment by bjorn3** [comment]
> Does C++ Itanium mangling allow for the `.llvm.` suffixes? AFAIK no, but you should get them if you use Clang with LTO.

https://itanium-cxx-abi.github.io/cxx-abi/abi.html#mangling-general

> Mangled names containing $ or . are reserved for private implementation use. Names produced using such extensions are inherently non-portable and should be given internal linkage where possible.

It does kind of allow them.

**Comment by nbdd0121** [comment]
> Does C++ Itanium mangling allow for the `.llvm.` suffixes? AFAIK no, but you should get them if you use Clang with LTO.

Yes.

>```
>    <mangled-name> ::= _Z <encoding>
>                   ::= _Z <encoding> . <vendor-specific suffix>
>    <encoding> ::= <function name> <bare-function-type>
>	       ::= <data name>
>	       ::= <special-name>
>```
>
> A `<mangled-name>` containing a period represents a vendor-specific version or portion of the entity named by the `<encoding>` prior to the first period. There is no restriction on the characters that may be used in the suffix following the period.

**Comment by nnethercote** [comment]
https://bugs.kde.org/show_bug.cgi?id=445916 has been filed for possibly updating Valgrind's v0 demangler to handle these suffixes, though it's still a bit unclear to me if that's the right thing to do.

**Comment by comex** [comment]
> though it's still a bit unclear to me if that's the right thing to do.

Well, if it helps:

With LLVM, names like `foo.llvm.3285396211802591752` and names like `foo.5` both arise when a symbol that originally had linkage local to a single object file, and thus only needed a locally unique name, goes through LTO and suddenly needs a globally unique name.

<details>
<summary>Details</summary>
<ul>
<li> Names like <code>foo.llvm.3285396211802591752</code> come from ThinLTO when a local symbol ends up being referenced from a different object file, because some reference to it that was originally from the first object file was inlined into the second one.  Since ThinLTO compiles each bitcode object file into its own native object file, such a symbol has to be changed to global linkage so that the native linker can resolve the reference; the suffix is preemptively attached to avoid clashes with any other symbols also named <code>foo</code> (which could be either a local symbol from another object file that was transformed in the same way, or a symbol that was already global).
<li> Names like <code>foo.5</code> come from full LTO.  Full LTO combines everything into a single native object file, so symbols that were local in the input bitcode object files can stay local in the output native object file, but the space of "local" names now includes all symbols from all of the input files.  Since full LTO is not incremental, the suffix is not attached preemptively, but only when there actually are two symbols with the same name from two different input files.
</ul>
</details>

In both cases, the suffixed `foo` has the same semantics as the original `foo`, and the suffix is just to ensure a unique name.

But there are also names like `foo.cold.1`, which represents a partial chunk of `foo` that was split off into its own function.  In this case, `foo.cold.1` does not have the same semantics as `foo`.  Ignoring the suffix is still fine if you are just trying to symbolicate a backtrace.  But for more obscure use cases it may not be fine.  Suppose you want to log whenever a function is called; treating each call to `foo.cold.1` as a call to `foo` would be misleading.

**Comment by comex** [comment]
Also, regarding the idea of just getting rid of the suffixes:

For Rust code, it might theoretically be possible to get rid of suffixes that exist to ensure global uniqueness, under the assumption that Rust compilers will never actually produce two unrelated symbols with the same mangled name.  But C and C++ compilers can and do produce such symbols (those languages expose local linkage directly in the form of `static`, so you just need `static` functions/variables with the same name in two different source files).  So Valgrind's C++ demangler at least ought to be dealing with these suffixes, and in theory the same code could be reused for Rust.

On top of that, even in Rust code, suffixes like `.cold.1` can't be avoided unless you want to get rid of the hot-cold splitting optimization altogether.  When LLVM splits out a chunk of a function into its own function, it has to give the new function *some* name.  It can't give it the same name as the original function since the original still exists.  In lieu of that, naming it after the original function plus a suffix is clearly more useful than, say, giving it a random name.

**Comment by str4d** [comment]
I'm writing a [Ghidra script for demangling Rust symbols](https://gist.github.com/str4d/e541f4c28e2bca80d222434ac1a204f4), and I have a question about the `<const>` grammar:
```
<const> = <type> <const-data>
        | "p" // placeholder, shown as _
        | <backref>
```

The `"p"` and `<backref>` cases are duplicated, because `<type>` is defined as:
```
<type> = <basic-type>
       | (... elided ...)
       | <backref>

<basic-type> = "a"      // i8
             | (... elided ...)
             | "p"      // placeholder (e.g. for generic params), shown as _
```

So it seems to me that `<const>` can be parsed as either `"p"` or `"p" <const-data>`, and similarly either `<backref>` or `<backref> <const-data>`. Is this intentional (i.e. is `<const-data>` partially-optional)? It seems like I need to do the following to process `<const>`:

- Parse `input` as `<type>`.
  - If `<type>` succeeds, parse `rest` with `<const-data>`.
    - If `<const-data>` succeeds, return `<type> <const-data>`.
    - If `<const-data>` fails, backtrack and inspect `<type>`.
      - If `<type> in ["p", <backref>]`, return `<type>`.
      - Else, `<const>` fails.
  - If `<type>` fails, then we know that `input` was not either `"p"` or `<backref>` (otherwise it would have succeeded), so `<const>` fails.

But I'm also not sure how to interpret a `<const>` with an optional `<const-data>`.

---

As an aside, this _feels_ ambiguous, because unlike other optional parts of the grammar, `<const-data>` has no distinguishing prefix:
```
<const-data> = ["n"] {<hex-digit>} "_"
```

I think it's actually unambiguous, but only indirectly, due to how sentinel characters of other parts of the grammar were selected.  Say we have `[[T; M]; N]`:
```
"A" "A" <type> <const> <const>
```
Given the current grammar, that could potentially be either:
```
"A" "A" <type> "p" {<hex-digit>} "_" <type> <const-data>
```
or:
```
"A" "A" <type> "p" <basic-type> <const-data>
```
and `<hex-digit>` can collide with `<basic-type>`.

However, there's nowhere else in the grammar that allows `<basic-type>` to be directly followed by `"_"`, so if the data contained `"p" <const-data>` but the parser tried to interpret `<const>` as `"p"` first, the parser will eventually figure out the problem and can backtrack.

I believe (but have not analyzed) that a collision in the other direction should also be detected indirectly. This partial optionality seems somewhat brittle though, especially as the RFC leaves extending `<const-data>` as a future task.

**Comment by str4d** [comment]
I had a look at how `rustc` implements const mangling:
https://github.com/rust-lang/rust/blob/028c6f1454787c068ff5117e9000a1de4fd98374/compiler/rustc_symbol_mangling/src/v0.rs#L578-L733

I think this corresponds to the following grammar:
```
<const> = "p"
        | <backref>
        | <subset-of-type> <const-data>
```

This also seems to be the grammar that [`rustc-demangle` uses](https://github.com/rust-lang/rustc-demangle/blob/2811a1ad6f7c8bead2ef3671e4fdc10de1553e96/src/v0.rs#L1082-L1209).

So I now believe this is a bug in the RFC.

**Comment by tmiasko** [comment]
@str4d in production `<const> = <type> <const-data>`, `<type>` is never `p`. There are pending updates to the grammar in https://github.com/rust-lang/rfcs/pull/3161 which make this explicit in the grammar.

**Comment by str4d** [comment]
Aha, yes that does indeed address my concern: `<const>` now never reaches `<type>` other than via `"V" <path> <const-fields>`, which has a separating prefix. Thanks!

**Comment by nnethercote** [comment]
https://bugs.kde.org/show_bug.cgi?id=445916 has details of some progress on the gcc/libiberty/valgrind side about handling the suffixes added by LLVM.

**Comment by eddyb** [comment]
@Amanieu According to the edit logs, almost exactly a year ago you checked the box for "Linux `perf`" ~~but nobody has changed this file at all for `v0`: [`torvalds/linux`@fb71c86 / `tools/perf/util/demangle-rust.c`](https://github.com/torvalds/linux/blob/5bfc75d92efd494db37f5c4c173d3639d4772966/tools/perf/util/demangle-rust.c)~~

I ran into lack of support while trying to get good symbol names with `perf record -g`.

<hr/>

*However*, looking closer at how it's driven, Rust `v0` support seems to "just" require `libbfd` from `binutils 2.36` (or later), so I'll add that to the checkbox, in case anyone else looks at it again (I happen to have `binutils 2.35.2` instead).

If anyone is familiar with Linux kernel patches, these can be removed nowadays (since Rust `legacy` demangling has been working through `libbfd` for a while now AFAICT):
* `tools/perf/util/demangle-rust.c`
* [the code invoking `demangle-rust.c` functions](https://github.com/torvalds/linux/blob/98f7fdced2e0efb1599a37b3e57671a7884f3a25/tools/perf/util/symbol-elf.c#L261-L266)

**Comment by Amanieu** [comment]
I checked the box for Linux perf because no change to the kernel was required: the new demangler will automatically get picked up from the updated libiberty.

**Comment by michaelwoerister** [comment]
FYI, @lqd opened a PR that extends the RFC with "vendor-specific suffixes" like `.llvm.123`: https://github.com/rust-lang/rfcs/pull/3224

Please provide your feedback if you have any.

**Comment by ojeda** [comment]
> If anyone is familiar with Linux kernel patches, these can be removed nowadays (since Rust `legacy` demangling has been working through `libbfd` for many years now AFAICT):
> 
> * [`tools/perf/util/demangle-rust.c`](https://github.com/torvalds/linux/blob/5bfc75d92efd494db37f5c4c173d3639d4772966/tools/perf/util/demangle-rust.c)
> * [the code invoking `demangle-rust.c` functions](https://github.com/torvalds/linux/blob/98f7fdced2e0efb1599a37b3e57671a7884f3a25/tools/perf/util/symbol-elf.c#L261-L266)

There was https://lore.kernel.org/lkml/20220201185054.1041917-1-german.gomez@arm.com/, but German notes:

> ```
> I have decided to drop this patch.
>
> It turns out that even shipped versions of libbfd and libiberty don't demangle some of the symbols completely
>
> For example:
>
> (doesn't strip away the hash at the end) _ZN10rs_tracing8internal11TRACE_STATE17h41dcd282cd61069dE.0                 ==> rs_tracing::internal::TRACE_STATE::h41dcd282cd61069d
> (doesn't demangle full symbol)           _ZN41_$LT$bool$u20$as$u20$core..fmt..Debug$GT$3fmt17h10f4b7b0094c3a75E.2262 ==> _$LT$bool$u20$as$u20$core..fmt..Debug$GT$::fmt::h10f4b7b0094c3a75
>
> These are cleaned up afterwards by perf's demangler.
> ```

**Comment by eddyb** [comment]
> It turns out that even shipped versions of libbfd and libiberty don't demangle some of the symbols completely

How is that possible? The code in `demangle-rust.c` is copy-pasted from what libiberty used to have for years (until I changed it when unifying it with the v0 demangler).

Also, I'm guessing the `.0` and `.2262` in those examples are stripped by `perf` before passing them off to libiberty? (since libiberty doesn't handle those suffixes correctly and just refuses to demangle entirely AFAIK - this is only now getting fixed)

Regarding the hash at the end, I think that's controlled by demangler flags (`-i` is `--no-verbose`):
```console
$ c++filt --version
GNU c++filt (GNU Binutils) 2.35.2
Copyright (C) 2020 Free Software Foundation, Inc.
This program is free software; you may redistribute it under the terms of
the GNU General Public License version 3 or (at your option) any later version.
This program has absolutely no warranty.
$ c++filt '_ZN10rs_tracing8internal11TRACE_STATE17h41dcd282cd61069dE'
rs_tracing::internal::TRACE_STATE::h41dcd282cd61069d
$ c++filt '_ZN41_$LT$bool$u20$as$u20$core..fmt..Debug$GT$3fmt17h10f4b7b0094c3a75E'
<bool as core::fmt::Debug>::fmt::h10f4b7b0094c3a75
$ c++filt -i '_ZN10rs_tracing8internal11TRACE_STATE17h41dcd282cd61069dE'
rs_tracing::internal::TRACE_STATE
$ c++filt -i '_ZN41_$LT$bool$u20$as$u20$core..fmt..Debug$GT$3fmt17h10f4b7b0094c3a75E'
<bool as core::fmt::Debug>::fmt
```

The lack of `$` unescaping is really worrying however - I can only reproduce if I *force* C++ demangling:
```console
$ c++filt --format gnu-v3 '_ZN41_$LT$bool$u20$as$u20$core..fmt..Debug$GT$3fmt17h10f4b7b0094c3a75E'
_$LT$bool$u20$as$u20$core..fmt..Debug$GT$::fmt::h10f4b7b0094c3a75
```

So maybe it's not the verbosity level, but "just" `perf` somehow forcing C++-only mode, instead of the automatic default?

Either that or *really old* libiberty versions, but the same code was added to both libiberty and `perf` around the same time.

**Comment by nnethercote** [comment]
> https://bugs.kde.org/show_bug.cgi?id=445916 has been filed for possibly updating Valgrind's v0 demangler to handle these suffixes, though it's still a bit unclear to me if that's the right thing to do.

The suffixes are now handled by gcc/libiberty, and those changes have been imported into Valgrind, and this Valgrind bug has been closed.

**Comment by bstrie** [comment]
Now that RFC https://github.com/rust-lang/rfcs/pull/3224 has been merged to resolve the suffix question, are there any implementations that still need updating in order to handle `.` and `$` suffixes? If not, shall we revive https://github.com/rust-lang/rust/pull/89917 for making v0 the default?

**Comment by programmerjake** [comment]
note that recent versions of rustc use `-C symbol-mangling-version=v0` rather than the `-Z` flag. the top comment led me astray..

**Comment by eddyb** [comment]
> note that recent versions of rustc use `-C symbol-mangling-version=v0` rather than the `-Z` flag. the top comment led me astray..

@programmerjake Thanks for bringing it up (it got forgotten) - I just tried updating it, is the new version better?

**Comment by programmerjake** [comment]
> > note that recent versions of rustc use `-C symbol-mangling-version=v0` rather than the `-Z` flag. the top comment led me astray..
> 
> @programmerjake Thanks for bringing it up (it got forgotten) - I just tried updating it, is the new version better?

yup, thx!

**Comment by eddyb** [comment]
So @Gankra was showing my some non-trivial `v0` symbols and that got me pondering about richer presentation than just text. [Even for plain text, I was able to prototype some stuff w/ `jq`](https://gist.github.com/eddyb/33c12bbf582fd9b5d93c8c2d3ed5c169), but that's a pile of hacks approximating "balanced `<>`/`()`" parsing.

Ideally we wouldn't be going the long way around through the "parse v0 mangling -> emit quasi-Rust (type) syntax -> parse quasi-Rust syntax -> pretty-print AST" pipeline.

I'm not sure what the status is on @michaelwoerister's [v0->AST demangler](https://github.com/michaelwoerister/std-mangle-rs) (which predates my "direct"/"allocation-less" v0 demangler in `rustc-demangle`) but from a quick glance it should be mostly compatible already? Additional testing and/or consolidation with the `rustc-demangle` repo would not be hard, if anyone is interested to pick it back up.
<sub>(Also, would it make sense to share code between them? Not trying to start a dozen bikesheds though, there should probably be dedicated issues for tracking anything that specific)</sub>

The trickier parts would be adding support for newer additions, e.g.:
* https://github.com/rust-lang/rustc-demangle/pull/55
  * <sub>OTOH, an AST demangler does have the advantage of being able to allocate e.g. the hex-encoded constant data payloads, and rely on standard APIs (`String::from_utf8`), instead of having to do everything on the fly like `rustc-demangler` does</sub>

<hr/>

How would tools even use a demangled AST to provide a better experience?
No one idea seems particularly strong on its own, but here's a few:
* minimal pretty-printing should be doable with little effort
  * <sub>e.g. try to fit each AST node on one line without going past terminal-width/80cols/some other limit, and once it fails for some node, all the "larger" nodes (that refer, directly or indirectly, to that one node) will be forced to a multi-line format (simply by e.g. building larger strings from smaller strings)</sub>
* the presence of backrefs can be exploited to cheaply refer to the same type many times
  * perhaps show "less important" occurrences in a more reduced form? (e.g. `…::Foo<…>` - literally using ellipses - instead of `foo::Foo<Bar, Baz>`)
* any interactive UI can fill in details outside-in, focusing on providing the most relevant information first (without permanently hiding anything, merely requiring additional interactions, like hovering over, or clicking an `…`)
* GUIs (and, to a lesser extent, TUIs) can vary color and other text aspects
  * <sub>syntax highlighting might be the first instinct, but other concepts like LISP-style "rainbow delimiters" (in our case, for any `<>`/`()`/`[]`/`{}`) or "semantic highlighting" (typically coloring identifiers based on their name resolution results, AFAICT), are probably more relevant</sub>
* <sub>I'm sure I forgot much cooler tricks, alas,</sub>

<hr/>

Anyway, all that said, I'm going to duplicate [some screenshots](https://twitter.com/eddyb_r/status/1552558087126618112) here as well:
1. HTML+CSS attempt at selective empahasis:
  <sup>(using a manually annotated demangling, with `-` standing in for `<small>` and `+` for `</small>`)</sup>
![image](https://user-images.githubusercontent.com/77424/181462089-08a3e591-2c5d-4763-81ed-3d65f2a9d7d7.png)
3. [cursed `jq` contraption](https://gist.github.com/eddyb/33c12bbf582fd9b5d93c8c2d3ed5c169) being applied to a longer Rust symbol <sup>(this one's automated; also works for C++)</sup>
![image](https://user-images.githubusercontent.com/77424/181446754-885e3c83-8dd4-4f52-b616-f59c3f9406af.png)

**Comment by EFanZh** [comment]
@eddyb I have written a demangler that can demangle v0 syntax: <https://github.com/EFanZh/ast-demangle/>.

**Comment by eddyb** [comment]
> @eddyb I have written a demangler that can demangle the latest v0 syntax symbol into a structured AST: https://github.com/EFanZh/ast-demangle/.

Oh, my bad @EFanZh, now that I'm looking at it, I'm pretty sure I've seen it before and just forgot :(

**Comment by pnkfelix** [comment]
Discussed in T-compiler [backlog bonanza](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bsteering.20meeting.5D.202022-08-12.20compiler-team.23484.20backlog.20b'za/near/293115939)

The v0 symbol mangling has been implemented. From https://github.com/rust-lang/rust/pull/89917 we have considered making v0 the default, but we have held off on doing so in order to give external tools time to add support. In PR #90054 we did make v0 the default for builds of `rustc` *itself* (but not object code generated by `rustc` on other programs).

We need to figure out what criteria we will use in this and other cases to decide that "it is time" to switch the defaults.

(We also considered opening a separate tracking issue for the question of "when to switch the default", but at this point I think we would only open such a tracking issue if we were ready to close this one, #60705, itself.

@rustbot label: S-tracking-needs-to-bake

**Comment by bstrie** [comment]
> we have held off on doing so in order to give external tools time to add support
>
> We need to figure out what criteria we will use in this and other cases to decide that "it is time" to switch the defaults.

The first thing to do would be to produce a list of tools that people want to support. For each tool, we should determine whether it supports v0, and, if so, the date of the first public release that features v0 support. Once each tool supports v0, and once each has supported v0 for long enough, then stabilization should be unblocked.

Obviously this list cannot guarantee that it will exhaustively mention every tool ever made, but the only alternative would be to never stabilize v0 for fear of overlooking some tool. In the meantime, we can use a blog post to put out a general call to tool developers to ask them to ensure that v0 works with their tools.

**Comment by jyn514** [comment]
> The first thing to do would be to produce a list of tools that people want to support. For each tool, we should determine whether it supports v0, and, if so, the date of the first public release that features v0 support. Once each tool supports v0, and once each has supported v0 for long enough (precise criteria TBD), then stabilization should be unblocked.
> 
> Obviously this list cannot guarantee that it will exhaustively mention every tool ever made, but the only alternative would be to never stabilize v0 for fear of overlooking some tool. In the meantime, we can use a blog post to put out a general call to tool developers to ask them to ensure that v0 works with their tools.

Nominating to hopefully act as a forcing function to create this list.

**Comment by nnethercote** [comment]
One problem with v0 mangling that hasn't been identified: it completely breaks the `cargo llvm-lines` tool. Here is example output with legacy mangling:
```
  Lines                 Copies              Function name
  -----                 ------              -------------
  134295                3225                (TOTAL)
    6102 (4.5%,  4.5%)    18 (0.6%,  0.6%)  alloc::raw_vec::RawVec<T,A>::grow_amortized
    2641 (2.0%,  6.5%)    64 (2.0%,  2.5%)  core::option::Option<T>::map
    2329 (1.7%,  8.2%)    17 (0.5%,  3.1%)  <core::slice::iter::Iter<T> as core::iter::traits::iterator::Iterator>::next
    1716 (1.3%,  9.5%)    11 (0.3%,  3.4%)  alloc::raw_vec::RawVec<T,A>::allocate_in
    1694 (1.3%, 10.8%)    15 (0.5%,  3.9%)  alloc::alloc::box_free
    1476 (1.1%, 11.9%)    18 (0.6%,  4.4%)  alloc::raw_vec::RawVec<T,A>::current_memory
    1461 (1.1%, 13.0%)     3 (0.1%,  4.5%)  hashbrown::raw::RawTable<T,A>::reserve_rehash
    1456 (1.1%, 14.1%)    16 (0.5%,  5.0%)  core::slice::iter::Iter<T>::new
    1249 (0.9%, 15.0%)     8 (0.2%,  5.3%)  <T as alloc::slice::hack::ConvertVec>::to_vec
    1065 (0.8%, 15.8%)     5 (0.2%,  5.4%)  aho_corasick::automaton::Automaton::leftmost_find_at_no_state_imp
```
And with v0 mangling:
```
  Lines                 Copies              Function name
  -----                 ------              -------------
  134295                3225                (TOTAL)
     960 (0.7%,  0.7%)     1 (0.0%,  0.0%)  <regex[455e3194582446bb]::prog::Program as core[d1a89b04220dd38d]::fmt::Debug>::fmt
     722 (0.5%,  1.3%)     1 (0.0%,  0.1%)  <regex[455e3194582446bb]::exec::ExecBuilder>::build
     544 (0.4%,  1.7%)     1 (0.0%,  0.1%)  <regex[455e3194582446bb]::dfa::Fsm>::exec_at
     497 (0.4%,  2.0%)     1 (0.0%,  0.1%)  <regex[455e3194582446bb]::compile::Compiler>::compile_many
     494 (0.4%,  2.4%)     1 (0.0%,  0.2%)  <aho_corasick[afd2d59d996825a5]::nfa::NFA<u32> as core[d1a89b04220dd38d]::fmt::Debug>::fmt
     487 (0.4%,  2.8%)     1 (0.0%,  0.2%)  <hashbrown[18cdbe82094945b3]::raw::RawTable<(&usize, &alloc[c687d6376d1d0c58]::string::String)>>::reserve_rehash::<hashbrown[18cdbe82094945b3]::map::make_hasher<&usize, &usize, &alloc[c687d6376d1d0c58]::string::String, std[e45faeee946555a1]::collections::hash::map::RandomState>::{closure#0}>
     487 (0.4%,  3.1%)     1 (0.0%,  0.2%)  <hashbrown[18cdbe82094945b3]::raw::RawTable<(alloc[c687d6376d1d0c58]::string::String, usize)>>::reserve_rehash::<hashbrown[18cdbe82094945b3]::map::make_hasher<alloc[c687d6376d1d0c58]::string::String, alloc[c687d6376d1d0c58]::string::String, usize, std[e45faeee946555a1]::collections::hash::map::RandomState>::{closure#0}>
     487 (0.4%,  3.5%)     1 (0.0%,  0.2%)  <hashbrown[18cdbe82094945b3]::raw::RawTable<(regex[455e3194582446bb]::dfa::State, u32)>>::reserve_rehash::<hashbrown[18cdbe82094945b3]::map::make_hasher<regex[455e3194582446bb]::dfa::State, regex[455e3194582446bb]::dfa::State, u32, std[e45faeee946555a1]::collections::hash::map::RandomState>::{closure#0}>
     456 (0.3%,  3.8%)     1 (0.0%,  0.3%)  <regex[455e3194582446bb]::compile::Compiler>::c_alternate
     433 (0.3%,  4.1%)     1 (0.0%,  0.3%)  <alloc[c687d6376d1d0c58]::alloc::Global as core[d1a89b04220dd38d]::alloc::Allocator>::shrink
```
Note the difference in the `copies` column. `cargo llvm-lines` entirely depends on the type-imprecison of legacy mangling. We go from having N different functions with the same name being combined, to every function being separate. E.g. with legacy mangling all the `grow_amortized` instances end up in the same bucket, while with v0 mangling they look like this:
```
     339 (0.3%,  6.8%)     1 (0.0%,  0.6%)  <alloc[c687d6376d1d0c58]::raw_vec::RawVec<(char, char)>>::grow_amortized
     339 (0.3%,  7.0%)     1 (0.0%,  0.6%)  <alloc[c687d6376d1d0c58]::raw_vec::RawVec<(u8, u32)>>::grow_amortized
     339 (0.3%,  7.3%)     1 (0.0%,  0.7%)  <alloc[c687d6376d1d0c58]::raw_vec::RawVec<(usize, usize)>>::grow_amortized
```
This is probably a case where `cargo llvm-lines` needs to change, rather than v0 mangling, but I thought it worth mentioning.

cc @dtolnay

**Comment by sanmai-NL** [comment]
Is or will there be an official Rust name mangling library (functionality), rather than demangling? Sometimes, one needs to mangle Rust item paths to look into binaries, e.g. like `perf` does. I hope there will be a reference implementation of specification.

**Comment by bjorn3** [comment]
Why would perf need to mangle names? There is no way to exactly reproduce symbol names outside of rustc itself given that they contain a crate disambiguator whose value depends on the `-Cmetadata` arguments passed when compiling the crate that defined the mentioned function/type (which for the standard library is unknown) as well as the exact rustc version used. Even two consecutive nightly releases will produce different symbol names.

**Comment by sanmai-NL** [comment]
Please re-read my sentence @bjorn3. I'm not claiming `perf` mangles names.

**Comment by sanmai-NL** [comment]
@bjorn3 Thanks for your explanation. I hope you didn't assume every commenter should know these details. I think the question is legitimate. [It was asked before in the context of GCC C++](https://stackoverflow.com/questions/41524956/gcc-c-name-mangling-reference). The information required I have available, but that's not important now. Even if just the algorithm were to be specified like in the GCC case, perhaps enough of the translation symbol to mangled symbol can be reconstructed to find the specific symbol in a binary for a given item path. That's my use case but I don't assume this would be the only solution of the only use case for a mangling spec or reference implementation.

**Comment by bjorn3** [comment]
> perhaps enough of the translation symbol to mangled symbol can be reconstructed to find the specific symbol in a binary for a given item path. That's my use case but I don't assume this would be the only solution of the only use case for a mangling spec or reference implementation.

It should be possible to have something like an api where you specify in the input a wildcard for the crate disambiguator and then the name mangling library would output a wildcard where it would otherwise print the crate disambiguator. Would this work for your use case?

**Comment by CAD97** [comment]
> I hope there will be a reference implementation of specification.

A reference implementation does already exist, essentially, as part of rustc. That it's not a reusable library just reflects that the goal of a known mangling scheme is the ability for 3rd party non-rustc tooling to be able to turn mangled symbols back into the demangled human-meaningful form. Being able to mangle symbols is explicitly a non-goal.

> Sometimes, one needs to mangle Rust item paths to look into binaries, e.g. like `perf` does.

For binary introspection, demangling is sufficient. Given an unmangled name, to find the corresponding mangled names[^s], you don't mangle the unmangled name to compare to the mangled symbols; instead, you demangle the symbols from the binary to compare to the unmangled symbol. Most of the time you'll want the full list of demangled symbols anyway, e.g. for display or otherwise.

[^s]: Name*s*, plural; multiple crates with the same name will have symbols which collide when unmangled and are disambiguated with the crate disambiguator.

If you want fully predictable names (e.g. for linking manually ABI-stable interfaces), you should be specifying them explicitly. It _would_ be interesting to be able to request v0 mangling (without the use of disambiguators) rather than having to manually apply a mangling scheme, but that's a completely separate feature request than the use for Rust-only names tracked here.

**Comment by sanmai-NL** [comment]
@CAD97 

When you need to step back to the same binary you demangled symbols, and determine to what mangled symbol a demangled name refers, then you may want this functionality. Please also consider that binaries you have and even a build pipeline including source code, does not mean you are free to modify the source code to achieve predicable symbol names or whatever.

By the way, it's bit of a semantic discussion what demangling entails, in response to my functional requirement at least. Third party tooling like `perf` may only demangle in a strict sense, but could be considered to mangle a given name that exists as a mangled symbol in the binary:

```sh
perf \
  probe \
    --exec $(realpath "mycrate/ target/debug/deps/binary-cfcd9bd03ac152c2") \
    --add="uprobe123=mycrate\:\:tests\:\:test_1"
```

Perf demangles the symbols and then matches with the unmangled name. So if it were to keep a mapping between the two, that would work as well. It may not amount to demangling in the general, but it would cover a use case.

**Comment by sanmai-NL** [comment]
> It should be possible to have something like an api where you specify in the input a wildcard for the crate disambiguator and then the name mangling library would output a wildcard where it would otherwise print the crate disambiguator. Would this work for your use case?

Yes, sure. And perhaps there are other, forensic cases and such. Please note, I'm not an expert or involved in this mangling work here or elsewhere, just chiming in as a user with a practical use case that I think will be relevant to a subgroup of real-world developers (not detailing it since it's part of a paper to be published).

**Comment by benpye** [comment]
> Now that RFC [rust-lang/rfcs#3224](https://github.com/rust-lang/rfcs/pull/3224) has been merged to resolve the suffix question, are there any implementations that still need updating in order to handle `.` and `$` suffixes? If not, shall we revive #89917 for making v0 the default?

As far as I can tell gdb does not support suffixes using `$` rather than `.`. It's also somewhat unfortunate that GDB strips the suffix, rather than including it in the demangled string - but that's not unbearable.

**Comment by bstrie** [comment]
@benpye Given that the only documented use of `$` suffixes in the wild is for thread-local data on Mach-O, I don't think it's a showstopper for shipping this as the default.

I wonder if the compiler team would like to use the upcoming 2024 edition as an excuse to finally ship v0 mangling? This would let us roll it out gradually and in a way that can be easily rolled back by users, and since it's an implementation detail we could make it the default for all editions someday in the future if we really wanted to.

**Comment by danobi** [comment]
As a heads up, `perf` is about to drop support for libbfd/libiberty for license incompatibility reasons. I think someone needs to go and teach perf how to demangle v0 symbols. More details here: https://lore.kernel.org/bpf/gnwmibvjtwboisw7uv32bdo4ziw4qzgwzvndqg2czpa6vp4olv@44n36ndbwobc/

**Comment by captain5050** [comment]
> As a heads up, `perf` is about to drop support for libbfd/libiberty for license incompatibility reasons. I think someone needs to go and teach perf how to demangle v0 symbols.

I sent:
https://lore.kernel.org/lkml/20250129193037.573431-1-irogers@google.com/
for this. There are some TODOs in it and it is only lightly tested.

**Comment by bjorn3** [comment]
Have you seen https://github.com/rust-lang/rustc-demangle/blob/main/crates/native-c/src/demangle.c? It is a C port of the official demangler with both v0 and legacy support.

**Comment by captain5050** [comment]
> Have you seen https://github.com/rust-lang/rustc-demangle/blob/main/crates/native-c/src/demangle.c? It is a C port of the official demangler with both v0 and legacy support.

I hadn't seen it. Given the official nature, legacy support and lack of TODOs compared to my version, I would say it was very much superior :-). Perhaps @arielb1 could contribute it to Linux perf? I'd be happy to try to co-develop if useful.

Wrt licenses, [MIT is explicitly compatible](https://git.kernel.org/pub/scm/linux/kernel/git/perf/perf-tools-next.git/tree/LICENSES/preferred/MIT?h=perf-tools-next) and [Apache 2 is compatible if combined with a compatible license under an OR license](https://git.kernel.org/pub/scm/linux/kernel/git/perf/perf-tools-next.git/tree/LICENSES/dual/Apache-2.0?h=perf-tools-next) which I believe agrees with the wording [here](https://github.com/rust-lang/rustc-demangle/blob/main/README.md) and is already in use by [other files](https://git.kernel.org/pub/scm/linux/kernel/git/perf/perf-tools-next.git/tree/rust/kernel/init.rs?h=perf-tools-next).

So outside of formatting things, I think the code can be added as is with the addition of a "SPDX-License-Identifier: Apache-2.0 OR MIT" comment at the top. There is a Linux kernel code contribution tool that will warn on formatting issues called [checkpatch.pl](https://git.kernel.org/pub/scm/linux/kernel/git/perf/perf-tools-next.git/tree/scripts/checkpatch.pl?h=perf-tools-next).

**Comment by captain5050** [comment]
Hi, wondered if there is any update? On the LKML thread it seems preferred to pull the code into the Linux tree rather than depend on a library (at least initially). If I did this I'm happy to add co-authored tags - I'm also happy to guide others to do the work. What's best? I'm happy to do the fairly trivial glue code. I worry given the need to stand up and make a decisions there's the potential for this issue to sit unresolved. Would love @arielb1's input given they wrote the code.

**Comment by lolbinarycat** [comment]
I un-checked the box for `perf` since it uses its own demangler (https://github.com/torvalds/linux/blob/master/tools/perf/util/demangle-rust.c) instead of deferring to a library as previously stated.

as bjorn mentioned, there is a C version of the demangler, it just needs someone to write up a patch to integrate it into perf.

**Comment by captain5050** [comment]
> as bjorn mentioned, there is a C version of the demangler, it just needs someone to write up a patch to integrate it into perf.

I wrote the patches and sent them to LKML, thanks to all and @arielb1 that the work builds from. Please let me know how I can improve the patches on the Linux mailing list:
https://lore.kernel.org/lkml/20250429185839.1807005-1-irogers@google.com/
For example, when adding the v0 demangling tests I switched to the alternate=true encoding as that better matched:
https://lore.kernel.org/lkml/20250429185839.1807005-5-irogers@google.com/
Thanks.

**Comment by arielb1** [comment]
Just noticed the issue now, sorry. Is there any help that is still needed?

If the question is about the alternate vs. non-alternate format - do whatever fits your case better. The alternate format is more verbose but preserves more information.

**Comment by captain5050** [comment]
The patches have been included into Linux v6.16:
https://lore.kernel.org/lkml/20250603203501.1961487-1-acme@kernel.org/

I was trying to go with the flow wrt licenses on tests and the like, hopefully the code, notes, attribution, tests.. all look good. You can see the source in Linus' tree here:
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/tools/perf/util/demangle-rust-v0.c
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/tools/perf/util/demangle-rust-v0.h
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/tools/perf/tests/demangle-rust-v0-test.c

Thanks!

**Comment by ReillyBrogan** [comment]
I noticed that this is now the default in the docs for 1.97 however it looks like the documentation was not updated to reflect that. There's no mention of it in `src/bootstrap/src/utils/change_tracker.rs` or the comment in `bootstrap.example.toml`.
